### PR TITLE
Don't insert duplicate account codes

### PR DESCRIPTION
**Pull request by kpp**
# Description

Check if account.code(hash) exists before inserts

### CODE REVIEW

File: crates/evm/src/evm/handler.rs [hunk 1]
```diff
@@ -557,9 +558,13 @@ fn calc_diff_size<EXT, DB: Database>(
             let account = &state[addr];
 
             if let Some(code) = account.info.code.as_ref() {
-                // if code is eoa code
-                diff_size += CODE_KEY_SIZE;
-                diff_size += code.len();
+                // Don't charge for account code if it is already in DB.
+                let db_code = db.code_by_hash(account.info.code_hash)?;
```
Comment (eyusufatik) [inline crates/evm/src/evm/handler.rs:562]: db is revm's internal cache. we have to check if !self.code.exists(code_hash, working_set)

Comment (kpp) [reply crates/evm/src/evm/handler.rs:562]: We don't have an access to self.code here

Comment (eyusufatik) [reply crates/evm/src/evm/handler.rs:562]: wait sorry, I got confused with journaled state. this is correct

File: crates/sovereign-sdk/module-system/sov-modules-core/src/storage/scratchpad.rs [hunk 2]
```diff
@@ -27,6 +27,9 @@ pub trait StateReaderAndWriter {
     /// Deletes a storage value.
     fn delete(&mut self, key: &StorageKey);
 
+    /// Checks a storage key for existance
```
Comment (eyusufatik) [inline crates/sovereign-sdk/module-system/sov-modules-core/src/storage/scratchpad.rs:30]:
```suggestion
    /// Checks a storage key for existence
```

File: crates/sovereign-sdk/module-system/sov-modules-core/src/storage/cache.rs [hunk 3]
```diff
@@ -419,6 +419,29 @@ impl StorageInternalCache {
         self.tx_cache.add_write(cache_key, None);
     }
 
+    /// Check a key for existance.
```
Comment (eyusufatik) [inline crates/sovereign-sdk/module-system/sov-modules-core/src/storage/cache.rs:422]:
```suggestion
    /// Check a key for existence.
```

**Review by eyusufatik — APPROVED** [review]

File: crates/evm/src/evm/handler.rs (hunk 1 above)
Comment (kpp) [inline crates/evm/src/evm/handler.rs:562]: Unfortunately there is no way to optimize this, we have to deser the code anyway. So there is no use in those .exists/ contains_key methods at all...

Comment (eyusufatik) [reply crates/evm/src/evm/handler.rs:562]: look, selfdestruct is very very rarely used. so it's not a big deal

File: crates/evm/src/evm/db_commit.rs [hunk 4]
```diff
@@ -51,8 +51,8 @@ impl<'a, C: sov_modules_api::Context> DatabaseCommit for EvmDb<'a, C> {
             let account_info = account.info;
 
             if let Some(ref code) = account_info.code {
-                if !code.is_empty() && !self.code.exists(&account_info.code_hash, self.working_set)
-                {
+                if !code.is_empty() {
+                    // TODO: would be good to have a contains_key method on the StateMap that would be optimized, so we can check the hash before storing the code
                     self.code
                         .set(&account_info.code_hash, code, self.working_set);
```
Comment (eyusufatik) [inline crates/evm/src/evm/db_commit.rs]: I wonder if we set the code state map to the same code again in a batch proof, do we include that in the state diff?

Comment (kpp) [reply crates/evm/src/evm/db_commit.rs]: No, it's gonna be only 1, because we keep only the last (k, v) pair.

Comment (eyusufatik) [reply crates/evm/src/evm/db_commit.rs]: no but imagine there is two different proofs 

proof 1: sequencer commitmenti in L1 #12
proof 2: sequencer commitment in L1 #15 

in proof one I deploy contract x 

state diff will have the code (as it should be)

in proof two I also deploy contract x
the value in the state tree does not change but will it be in the state diff????

Comment (eyusufatik) [reply crates/evm/src/evm/db_commit.rs]: it's ok. sorry for the misunderstanding. we have the if check now

**Pull request merged**